### PR TITLE
Helm: fixed operator extra-args

### PR DESCRIPTION
**Pull request by sergeimonakhov**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

This fix, adds extra-args for the cilium-operator to the helm chart

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (nbusseneau) [inline install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: Oooh, this is a clever approach! I suppose `extraArgs` are then setup with something like this?

```yaml
operator:
    extraArgs:
        option1: value1
        option2: value2
        flag1:
        flag2:
```

It looks neat, but I would tend to think using a simple array would be more robust:

```yaml
operator:
    extraArgs:
        - --option1=value1
        - --option2=value2
        - --flag1
        - -f2
```

Which would be inserted with something like above for annotations:

```go
{{- with .Values.operator.extraArgs }}
        {{- toYaml . | nindent 8 }}
{{- end }}
```

In particular, this would allow usage of any arbitrary command line arguments (e.g. in short syntax `-v`), and not only those formatted as expected by `--key=value` or `--key`. I think it is more flexible :)

What do you and other reviewers think?

**Review by nbusseneau — CHANGES_REQUESTED** [review]
Thanks for the PR!

I have reviewed and commented the template modification below. After reviewing these and making a decision, we will also have to update:

- [ ] Documentation in `install/kubernetes/cilium/README.md`
- [ ] Default values in `install/kubernetes/cilium/values.yaml`

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (sergeimonakhov) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: All other templates use the code above...

Comment (nbusseneau) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: Sorry, I am new :D You are right: I missed that there was one instance of `extraArgs` already set up this way in `cilium-agent-daemonset.yaml`. That said, it seems to be the only one...? The other `extraArgs` values are seemingly never used either, so perhaps we should take advantage of this PR to standardize and retrofit to all `extraArgs` instances, whatever decision is made re: object vs. array 🤔 

I will wait for other reviewers to chime in on this point.

Comment (nebril) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: Agreed with @Skymirrh here.

**Review by nebril — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can you put some more information in your commit message? At the very least indicating that it's about helm would be a good thing to have there.

**Comment by D1abloRUS** [comment]
> Thanks for the PR!
> 
> Can you put some more information in your commit message? At the very least indicating that it's about helm would be a good thing to have there.

Updated

**Comment by qmonnet** [comment]
> Updated

Thanks! Can you please also address https://github.com/cilium/cilium/pull/14974#discussion_r576284487?

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (sergeimonakhov) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: what will we use as a result? `toYaml`?

Comment (nbusseneau) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: I'm waiting @seanmwinn's sentiment on the topic.

Also, I would like to semi-retract something:

> perhaps we should take advantage of this PR to standardize and retrofit to all `extraArgs` instances

This was poorly conveyed: we should not expect you to do that refactoring. Of course if you feel like refactoring all instances, you're more than welcome, but if you don't, then that's also OK. All contributions are welcome :)

Comment (errordeveloper) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: I personally think that simple arrays is indeed a better options, but given that the proposed change is based on code pattern that existed elsewhere, I think it's reasonable to merge this PR as it stands :)

**Review by errordeveloper — APPROVED** [review]

**Comment by joestringer** [comment]
Hi, are next steps for this PR clear for @D1abloRUS here? I can't quite tell whether we're waiting on feedback to be addressed or trying to come to a conclusion on the discussion thread here: https://github.com/cilium/cilium/pull/14974#discussion_r576213685 . I've pinged @seanmwinn privately, but also currently the PR is blocked by reviews from @nebril @Skymirrh . Just want to check that we're all on the same page on how to move this forward.

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:87]: I'm fine with whatever is going to be standardized across the entire chart. There are arguments for/against using either pattern. Feedback here indicates users are more in favor of using an array, which is fine.

I would just add that extraArgs is traditionally a development-cycle flag. It is useful for injecting flags for a binary at runtime. If we expect users to be able to tune or configure a setting regularly, I would suggest exposing a Helm value to make it tunable instead of relying on a long list of extraArgs.

**Review by seanmwinn — APPROVED** [review]

**Comment by Skymirrh** [comment]
@joestringer The problem I have right now is we have a threefold situation with conflicting resolutions, which I believe requires a consensus:

- Inconsistencies in the current state of the chart:
  - `extraArgs` are documented (in `README`) as available for `agent` (top-level `extraArgs` in `values.yaml`), `etcd`, `nodeinit`, `operator` and `preflight` (all the latter in their respective component subsection). But in reality, only `agent` actually uses `extraArgs`, all the other ones are unused. I suspect this situation to be a leftover from a refactoring. Whatever the case, we probably want to fix this discrepancy (either by removing documentation for unused values, or adding in the corresponding template inclusions).
  - The only `extraArgs` currently implemented (for `agent`) uses an object-based pattern, is already out in the wild, and thus is potentially used by chart users. We probably do not want to change its type.
- Argument around object-based `extraArgs` vs. array-based:
  - Array-based is more flexible in usage terms (allows for arbitrary arguments) and more easily included in templates (simple `toYaml`), but less elegant/readable.
  - Object-based is more elegant in YAML terms (more readable/parsable), but less flexible (limited to `--key=value`/`--key`-formatted arguments, does not allow multiple arguments with same `key`) and requires a bit of boilerplate for inclusion in templates.
  - Both are equivalently standard in Helm charts. I've checked a bunch of community Helm charts and both are used extensively. Just for illustration, here's [Kubernetes Dashboard](https://github.com/kubernetes/dashboard/search?q=extraArgs) with array-based, [Kubernetes NGINX Ingress Controller](https://github.com/kubernetes/ingress-nginx/search?q=extraArgs) with object-based, [Prometheus](https://github.com/prometheus-community/helm-charts/search?q=extraArgs) with both.
- We currently have two parallel PRs dabbling around this discussion:
  - #14974 (this one), which applies the object-based approach to one of the missing documented `operator.extraArgs`.
  - #14947, which adds the array-based approach as `extraArgsList` to the existing object-based `extraArgs` capabilities for `agent`, specifically because of the  inflexibility of `extraArgs`).

I believe we should make a decision before closing/merging either one, or merging both, depending on which scenario we   want to support going forward:

1. Keep object-based approach
    - Close #14947
    - Merge #14974
    - Decide which documented `*.extraArgs` should actually be offered or not, and remove documentation/implement accordingly
2. Switch to array-based approach
      - Close #14974
      - Deprecate `extraArgs` but keep it in the code, remove documented `*.extraArgs`
      - Decide how to offer `extraArgs` as array-based (I like `extraArgsList` and `extraArgsArray`), for `agent` (without masking the deprecated `extraArgs`) and for other components (with `extraArgs`, even though type is different? Or with the same name as `agent`? I prefer the latter)
      - Merge #14947 once modified accordingly
3. Support both approaches
      - Merge #14974
      - Decide which documented `*.extraArgs` should actually be offered or not, and remove documentation/implement accordingly
      - Decide how to offer `extraArgsList` (or whatever name) on top of `extraArgs`
      - Merge #14947 once modified accordingly

**Comment by joestringer** [comment]
Thanks for laying it out @Skymirrh, the summary helped to get me up to speed. Only thing I wish is that Github didn't allocate such similar PR numbers for the two PRs we're discussing here :wink: 

Given that there's already PR #14947 out suggesting that the object-based approach doesn't work for some use cases (specifically, when you need to pass the same flag multiple times), I'd be inclined to say that we need to introduce an array-based approach either way, so that would remove option (1).

I see ExtraArgs as kinda a "convenience feature" for cases where the charts don't explicitly provide an option to customize what you want, then you can just specify it there. The majority of users should not be using it, and if users are using it then hopefully they're thinking about contributing back core changes to the helm charts to expose the knobs they really need. From that perspective I feel like as long as we properly document how the ExtraArgs is changing from v1.9 to v1.10 (including examples of what you might have done in v1.9 and how you achieve the same in v1.10), then dropping the object-based approach is OK to me.

Supporting both approaches doesn't seem strictly necessary to me, and having two different syntaxes to support similar use cases seems likely to cause additional confusion for exactly how to pass the parameters depending on which field name you use. Sure the object syntax is nicer but if we want nice syntax, we can improve the core helm charts themselves.

If we're worried about the upgrades affecting existing users, then perhaps we could use a different name for the list-based approach (like #14947 does today) and warn users who try to use the object-based approach with a nice helm warning saying "use this approach instead, for more info see this docs page".

**Comment by Skymirrh** [comment]
Agree with all your points. So, I take it decision is to be made between these two scenarios:

1. Deprecate
    - Deprecate object-based `extraArgs`, add array-based `extraArgsList` (or whatever name)
    - Document Helm chart depreciation and recommend switching
2. Replace
    - Replace object-based `extraArgs` with array-based `extraArgs`
    - Document Helm chart breaking change.

I personally am biased towards 2, but this is just because it looks more elegant, so it's not really an argument.

No matter the choice, afterwards we should remember to add the array-based solution where desirable (I suppose in all modules, to make it consistent and practical since it's more of a development tool).

**Comment by Skymirrh** [comment]
Recap from community meeting:
- `extraArgs` was actually introduced recently (in 1.9) and is totally intended for CI / developer testing, not end users / customers (thanks @seanmwinn for the history).
- This means we can break stuff.

=> looks like this supports the second approach:
- Replace object-based extraArgs with array-based extraArgs
- Document Helm chart breaking change
- Implement array-based solution for all components

I've opened issue #15150 for tracking this.

**Comment by Skymirrh** [comment]
Since goal has changed and is now tracked in #15150, I will be closing this PR.

@D1abloRUS I would like to thank you again for your contribution! Even if the PR was not merged in the end, it was plenty useful: it shed light on documentation inconsistencies and improvements to be made on Helm charts, and thanks to the conversation you sparkled we now have a clear idea of how to fix the situation :) You are of course most welcome to continue contributing, whether for #15150 or anything you wish to.

**Pull request closed without merging**
